### DISCUSSION TITLE
More granular IR format

### DISCUSSION THREAD

**Discussion by AttilaMihaly** [finos/morphir#203] [Ideas]
Morphir tools currently communicate through a single JSON file that contains the IR for all the modules within a package and also the package's dependencies. While this is a simple approach that works the community brought up the idea of creating a more granular format that breaks up the IR into smaller files. 

We had a few discussions about the implementation details and it seems to be feasible but to come up with the best design we need spell-out specific requirements. So in this discussion I would like to invite the community to share what specific advantages we are hoping to get from a more granular IR structure.

To get things started, I know of one particular advantage:
 - If we decide to put the IR in source control a granular format will yield better diffs in PRs.

What other benefit are we hoping to achieve?
@DamianReeves, @stephengoldbaum

**Comment by stephengoldbaum** [comment]
I have a bunch of thoughts on this. Rather than wait until I've collected them all, I'll drip them in.  My main hypothesis is that **a file-based format will be more accessible for new developers and tool writers**.  Some thoughts:

- **_Easy traversing_** - Most developers are comfortable working with the file system and there are a wealth of tools across many technologies.  For example, want to find all of the Types in a distribution?  Easy enough to do just with `UNIX find`: `find -name "*.type"`.  
- **_Maps naturally to URLs and REST_** - That should make it easy to write web UIs and REST tools.
- **_Modularizable_** - We can extend the IR by adding new file types alongside the core files.  This would work really well for decorations.  For example, if we have a type, `Foo`, defined in `Foo.type` and a decoration for data sensitivity, we can simply add `Foo.sensitivity` alongside it and it will get packaged up with the distribution.
- **_Diffing across versions_** - It's difficult to see what has changed across versions in the single-file IR with standard diff tools (like git). Having separate files makes it much easier with those tools.

Those are my main points for now.
